Create hex image

alekob opened the pull request:
Create an extension to "generatebits" in order to write a hex image data and / or instruction memories. This enables the user to use one single (e. g. ROM) image for ASIC simulation and FPGA prototyping.

pjaaskel commented on tce/src/applibs/PIG/AsciiImageWriter.cc:
```diff
@@ -138,3 +144,62 @@ AsciiImageWriter::writeSequence(
+    int lastIndex = nextBitIndex_ + length - 1;
+
+    if (lastIndex >= bits_.size() && !padEnd) {
+        const string procName = "AsciiImageWriter::writeHexSequence";
```
You can replace this with \_\_func\_\_.

pjaaskel commented on tce/src/applibs/PIG/AsciiImageWriter.cc:
```diff
@@ -138,3 +144,62 @@ AsciiImageWriter::writeSequence(
+    std::vector<bool> bitRow;
+
+    // Copy bitstream for one row
+    for(int i=nextBitIndex_; i<=lastIndex; ++i) {
```
Plase add space after 'for'

pjaaskel commented on tce/src/applibs/PIG/AsciiImageWriter.cc:
```diff
@@ -138,3 +144,62 @@ AsciiImageWriter::writeSequence(
+    }
+
+    // optionally extend bit stream
+    int numPadBits = 4*nibbleCount - length;
```
Please add space after and before *

pjaaskel commented on tce/src/applibs/PIG/AsciiImageWriter.cc:
```diff
@@ -138,3 +144,62 @@ AsciiImageWriter::writeSequence(
+
+    // Generate a list of nibble
+    std::vector<uint8_t> Nibble;
+    for (std::vector<bool>::iterator it=bitRow.begin(); it<bitRow.end(); it+=4) {
```
Ditto before/after =.

pjaaskel commented on tce/src/applibs/PIG/AsciiImageWriter.cc:
```diff
@@ -138,3 +144,62 @@ AsciiImageWriter::writeSequence(
+    }
+
+    // "print" hex stream
+    for (std::vector<uint8_t>::iterator ui8=Nibble.begin(); ui8<Nibble.end(); ++ui8) {
```
White spaces needed here. Also check that lines are below 80 chars of length.

pjaaskel commented on tce/src/applibs/PIG/AsciiImageWriter.hh:
```diff
@@ -34,6 +34,9 @@
 #ifndef TTA_ASCII_IMAGE_WRITER_HH
 #define TTA_ASCII_IMAGE_WRITER_HH
 
+#include <cstdint>
```
If not needed in the header, please move the includes to the .cc to reduce cpp effort.

pjaaskel commented on tce/src/applibs/PIG/HexImageWriter.cc:
```diff
@@ -0,0 +1,90 @@
+/*
+    Copyright (c) 2002-2018 Tampere University of Technology.
```
Also add your name to the copyright as you own the code (no copyright assignments to TUT to my knowledge). Tampere University of Technology and [YOURNAME] is probably fair as it uses TUT written interfaces.

pjaaskel commented on tce/src/applibs/PIG/HexImageWriter.hh:
```diff
@@ -0,0 +1,52 @@
+/*
+    Copyright (c) 2002-2018 Tampere University of Technology.
+
+    This file is part of TTA-Based Codesign Environment (TCE).
```
Here too.

pjaaskel commented:
Info for developers: http://tce.cs.tut.fi/cgi-bin/moin.cgi/Development the testing guide: http://tce.cs.tut.fi/cgi-bin/moin.cgi/TestSuite

tervoa commented:
Program image generation with an instruction width not divisible by 4 (e.g. `minimal.adf`) fails with the following:
```
terminate called after throwing an instance of 'std::length_error'
  what():  vector<bool>::_M_insert_range
Aborted
```

alekob commented:
Thanks for all of your helpful comments and testings. (Hopefully) I've fixed all issues and added a new test in my latest commit a2c4fb9802e3383d4bfb5632ec4cd182180abea2
